### DISCUSSION TITLE
ERROR 403: project not open or last run not ended

### DISCUSSION THREAD

**Discussion by shizhiman** [pyswmm/pyswmm#351] [Q&A]
Hi everyone!
I am quite new to pyswmm. For study purposes, I am supposed to use pyswmm to test multi-threaded simulation.
However, I encountered this Error 403:project not open or last run not ended.I use two method,using step,it not work.
so I want to know: when multi-threaded simulation using step,it not work? How to multi-threaded simulation?
Thank you very much in advance!

The development environment:
OS: Windows 7 SP1
pyswmm package version: 1.1.1
swmm-toolkit: 0.8.2
Python version (64 bit): Python3.6.5 

Test Code:

`# author: Zhiman Shi
# create time :2022/7/4 16:08
import time
import datetime
from pyswmm import Simulation
import threading
import traceback
from threading import Thread

class ParalSimulation():
    def simulation_rainfall(self,file=None,test_type=None):
        test_time_start = datetime.datetime.now()
        test =1
        # method two:  It not work ================================================================
        sim = Simulation(file)
        try:
            for step in sim:
                if test ==1:
                    time.sleep(2)
                elif test ==2:
                    time.sleep(2)
                test = test + 1
            # sim.report()
            sim.close()
        except Exception as e:
            print(traceback.format_exc())
        test_time_end = datetime.datetime.now()
        duration = ( test_time_end - test_time_start).seconds
        print(f'simulation-【{test_type}】-运行时长= {duration} 秒。开始时间【{test_time_start}】-结束时间【{test_time_end}】')
        # method one: This method  ok ================================================================
        # sim = Simulation(file)
        # time.sleep(8)
        # sim.execute()

if __name__ == '__main__':
    a_INP_file  =r'C:\Users\dell\Desktop\WZPS\testaysnc\dl_template_a.inp'
    b_INP_file  =r'C:\Users\dell\Desktop\WZPS\testaysnc\dl_template_b.inp'

    t1 = threading.Thread(target=ParalSimulation().simulation_rainfall, args=(a_INP_file,'First'))
    t2 = threading.Thread(target=ParalSimulation().simulation_rainfall, args=(b_INP_file,'Second'))
    t1.start()
    t2.start()

    t2.join()
    t1.join()
`
                                                                                                         Amanda

**Comment by bemcdonnell** [comment]
@shizhiman,

So the way that SWMM is architected precludes you from being able to open multiple models in one “instance” of Python.  Therefore, if you want to run optimization in multiple threads, you need to use Threading AND subprocess.  Subprocess will simply be used to execute a command line operation where you will have “python pyswmm_code.py <args>“.  I might have some example codes you can look at but unfortunately I am away from my desk today. Let me try for tomorrow.  I suggest you take a look at how our friends did this in OSTRICH_SWMM (@kmmacro)!

**Reply by shizhiman** [reply]
@bemcdonnell，
Thank you very much for your timely answer.  I am sorry to disturb you while you are resting. I look at how our friends did this in OSTRICH_SWMM firstly (@kmmacro), and I look forward to seeing your example codes. Could you please send the example to my email (szhiman@126.com) when it is convenient for you? Thank you again!
